src/backend/distributed/transaction/transaction_management.c

```diff
@@ -801,3 +864,41 @@ TriggerMetadataSyncOnCommit(void)
 {
 	MetadataSyncOnCommit = true;
 }
+
+
+/*
+ * Function raises an exception, if the current backend started a coordinated
+ * transaction and got a PREPARE event to become a participant in a 2PC
+ * transaction coordinated by another node.
+ */
+static void
+EnsureSafe2PC(void)
+{
+	if (!InCoordinatedTransaction())
+	{
+		/* If the backend has not started a coordinated transaction */
+		return;
+	}
+
+	if (GetInForceDelegatedFuncExecution())
+	{
+		/*
+		 * If a distributed forcePushdown function is executed in the
+		 * current coordinated transaction, allow restricted 2PC.
+		 */
+		return;
+	}
+
+	if (IsCitusInitiatedRemoteBackend() && EnableUnsafeTriggers)
```

Comment:
Not thrilled about the usage of EnableUnsafeTriggers on the workers here. This gives the GUC a lot of hidden side-effects.

e.g. does the user know that the setting needs to be applied permanently on the workers for triggers to be fully functional?

e.g. what if one user created a trigger and enables this GUC and everything works fine, but then another user tries to delete and it breaks

I'd suggest only checking IsCitusInitiatedRemoteBackend()

Reply:
Removed it, was trying to be conservative.